File: modules/spread/app/model/spread/ODFCorrector.java
package model.spread;

import java.io.File;
import java.io.FileOutputStream;
import java.nio.file.Path;
import java.util.ArrayList;

import org.odftoolkit.odfdom.type.Color;
import org.odftoolkit.simple.SpreadsheetDocument;
import org.odftoolkit.simple.style.Font;
import org.odftoolkit.simple.style.StyleTypeDefinitions.FontStyle;
import org.odftoolkit.simple.table.Cell;
import org.odftoolkit.simple.table.Table;

public class ODFCorrector extends SpreadCorrector<SpreadsheetDocument, Table, Cell, Font, Color> {
  
  // TODO: magic numbers...
  private static final int MAXROW = 80;
  private static final int MAXCOLUMN = 22;
  
  private static final String COLOR_WHITE = "#FFFFFF";
  private static final String FONT = "Arial";
  private static final double FONT_SIZE = 10.;
  
  @Override
  protected void closeDocument(SpreadsheetDocument document) {
    document.close();
  }
  
  @Override
  protected String compareCellFormulas(Cell masterCell, Cell compareCell) {
    String masterFormula = masterCell.getFormula();
    String compareFormula = compareCell.getFormula();
    if(masterFormula == null)
      // Keine Formel zu vergleichen
      return "";
    else if(masterFormula.equals(compareFormula))
      // Formel richtig
      return "Formel richtig.";
    else if(compareFormula == null)
      // Keine Formel von Student angegeben
      return "Keine Formel angegeben!";
    else {
      String diffOfTwoFormulas = HashSetHelper.getDiffOfTwoFormulas(masterFormula, compareFormula);
      if(diffOfTwoFormulas.equals(""))
        // TODO: can this happen?
        return "Formel richtig.";
      else
        return "Formel falsch. " + diffOfTwoFormulas;
    }
  }
  
  @Override
  protected String compareCellValues(Cell masterCell, Cell compareCell) {
    String masterValue = masterCell.getStringValue(), compareValue = compareCell.getStringValue();
    // FIXME: why substring from 0 to first newline?
    if(compareValue.indexOf("\n") != -1)
      compareValue = compareValue.substring(0, compareValue.indexOf("\n"));
    if(compareValue.isEmpty())
      return "Keinen Wert angegeben!";
    else if(masterValue.equals(compareValue))
      return "Wert richtig.";
    else
      return "Wert falsch. Erwartet wurde '" + masterValue + "'.";
  }
  
  @Override
  protected String compareChartsInSheet(Table compareSheet, Table sampleSheet) {
    // FIXME: nicht von ODFToolkit unterstützt...
    return null;
  }
  
  @Override
  protected String compareNumberOfChartsInDocument(SpreadsheetDocument compare, SpreadsheetDocument sample) {
    int sampleCount = sample.getChartCount(), compareCount = compare.getChartCount();
    if(sampleCount == 0)
      return "Es waren keine Diagramme zu erstellen.";
    else if(sampleCount != compareCount)
      return "Falsche Anzahl Diagramme im Dokument (erwartet: " + sampleCount + ", gezählt: " + compareCount + ").";
    else
      return "Richtige Anzahl Diagramme gefunden.";
  }
  
  @Override
  protected void compareSheet(Table sampleTable, Table compareTable, boolean correctConditionalFormating) {
    if(correctConditionalFormating) {
      // NOTICE: Does not work in ODF Toolkit
    }
    // Iterate over colored cells
    ArrayList<Cell> range = getColoredRange(sampleTable);
    for(Cell cellMaster: range) {
      int rowIndex = cellMaster.getRowIndex();
      int columnIndex = cellMaster.getColumnIndex();
      Cell cellCompare = compareTable.getCellByPosition(columnIndex, rowIndex);
      
      if(cellCompare == null)
        // TODO: Fehler werfen? Kann das überhaupt passieren?
        return;
      
      // Compare cell values
      String cellValueResult = compareCellValues(cellMaster, cellCompare);
      String cellFormulaResult = compareCellFormulas(cellMaster, cellCompare);
      
      setCellComment(cellCompare, cellValueResult + "\n" + cellFormulaResult);
      
      if(cellValueResult.equals("Wert richtig.")
          && (cellFormulaResult.isEmpty() || cellFormulaResult.equals("Formel richtig.")))
        setCellStyle(cellCompare, new Font(FONT, FontStyle.BOLD, FONT_SIZE), Color.GREEN);
      // cellCompare.setFont(new Font(FONT, FontStyle.BOLD, FONT_SIZE,
      // Color.GREEN));
      else
        setCellStyle(cellCompare, new Font(FONT, FontStyle.ITALIC, FONT_SIZE), Color.RED);
      // cellCompare.setFont(new Font(FONT, FontStyle.ITALIC, FONT_SIZE,
      // Color.RED));
    }
  }
  
  @Override
  protected Cell getCellByPosition(Table table, int column, int row) {
    return table.getCellByPosition(column, row);
  }
  
  @Override
  @SuppressWarnings("deprecation")
  protected ArrayList<Cell> getColoredRange(Table master) {
    ArrayList<Cell> range = new ArrayList<Cell>();
    for(int row = 0; row < MAXROW; row++) {
      for(int column = 0; column < MAXCOLUMN; column++) {
        Cell oCell = master.getRowByIndex(row).getCellByIndex(column);
        if(!oCell.getCellBackgroundColorString().equals(COLOR_WHITE))
          range.add(oCell);
      }
    }
    return range;
  }
  
  @Override
  protected Table getSheetByIndex(SpreadsheetDocument document, int sheetIndex) {
    return document.getSheetByIndex(sheetIndex);
  }
  
  @Override
  protected int getSheetCount(SpreadsheetDocument document) {
    return document.getSheetCount();
  }
  
  @Override
  protected SpreadsheetDocument loadDocument(Path path) {
    try {
      return SpreadsheetDocument.loadDocument(path.toFile());
    } catch (Exception e) {
      return null;
    }
  }
  
  @Override
  protected void saveCorrectedSpreadsheet(SpreadsheetDocument document, Path testPath) {
    // TODO userFolder: saveFolder!
    String userFolder = SpreadSheetCorrector.getUserFolder(testPath);
    String fileName = SpreadSheetCorrector.getFileName(testPath);
    try {
      // FIXME: use document.save(File file);
      // File saveTo = new File("TODO");
      File dir = new File(userFolder);
      if(!dir.exists()) {
        dir.mkdirs();
      }
      FileOutputStream fileOut = new FileOutputStream(userFolder + fileName + "_Korrektur.ods");
      document.save(fileOut);
      fileOut.close();
    } catch (Exception e) {
      System.out.println(e);
    }
  }
  
  @Override
  protected void setCellComment(Cell cell, String message) {
    if(message == null || message.isEmpty())
      return;
    cell.setNoteText(message);
  }
  
  @Override
  protected void setCellStyle(Cell cell, Font font, Color color) {
    font.setColor(color);
    cell.setFont(font);
  }
  
}

File: app/model/spread/ODFCorrector.scala
package model.spread

import java.io.FileOutputStream
import java.nio.file.{Files, Path, Paths}

import com.google.common.io.{Files => GFiles}
import model.spread.SpreadConsts._
import model.spread.SpreadUtils._
import org.odftoolkit.odfdom.`type`.Color
import org.odftoolkit.simple.SpreadsheetDocument
import org.odftoolkit.simple.style.Font
import org.odftoolkit.simple.style.StyleTypeDefinitions.FontStyle
import org.odftoolkit.simple.table.{Cell, Table}

import scala.util.Try

object ODFCorrector extends SpreadCorrector[SpreadsheetDocument, Table, Cell, Font, Color](Color.GREEN, Color.RED) {

  val maxRow    = 80
  val maxColumn = 22

  val colorWhite       = "#FFFFFF"
  val fontName         = "Arial"
  val fontSize: Double = 10

  override def closeDocument(document: SpreadsheetDocument): Unit = document.close()

  override def compareCellFormulas(masterCell: Cell, compareCell: Cell): (Boolean, String) = Option(masterCell.getFormula) match {
    case None                => (true, noFormulaRequired)
    case Some(masterFormula) => Option(compareCell.getFormula) match {
      case None                 => (false, formulaMissing)
      case Some(compareFormula) =>
        if (masterFormula == compareFormula) (true, formulaCorrect)
        else {
          val diffOfTwoFormulas = SpreadUtils.getDiffOfTwoFormulas(masterFormula, compareFormula)
          if (diffOfTwoFormulas.isEmpty) (true, formulaCorrect)
          else (false, s"Formel falsch. $diffOfTwoFormulas")
        }
    }
  }

  override def compareCellValues(masterCell: Cell, compareCell: Cell): (Boolean, String) = {
    val masterValue = masterCell.getStringValue
    val compareCellValue = compareCell.getStringValue
    val compareValue = if (compareCellValue.contains('\n')) compareCellValue.substring(0, compareCellValue.indexOf('\n')) else compareCellValue

    if (compareValue.isEmpty) (false, COMMENT_VALUE_MISSING)
    else if (masterValue == compareValue) (true, COMMENT_VALUE_CORRECT)
    else (false, s"Wert falsch. Erwartet wurde '$masterValue'.")
  }

  override def compareChartsInSheet(compareSheet: Table, sampleSheet: Table): (Boolean, String) = (false, "") // FIXME: nicht von ODFToolkit unterstützt...

  override def compareNumberOfChartsInDocument(compare: SpreadsheetDocument, sample: SpreadsheetDocument): (Boolean, String) =
    sample.getChartCount match {
      case 0           => (false, COMMENT_CHART_FALSE)
      case sampleCount =>
        val compareCount = compare.getChartCount
        if (sampleCount == compareCount) (true, COMMENT_CHART_NUM_CORRECT)
        else (false, s"Falsche Anzahl Diagramme im Dokument (Erwartet: $sampleCount, Gefunden: $compareCount).")
    }

  override def compareSheet(sampleTable: Table, compareTable: Table, conditionalFormating: Boolean): Unit = {

    if (conditionalFormating) {
      val conditionalFormattingResult = compareSheetConditionalFormatting(sampleTable, compareTable)
      getCellByPosition(compareTable, 0, 0) match {
        case None       => Unit
        case Some(cell) => setCellComment(cell, conditionalFormattingResult.mkString("\n"))
      }
    }

    for (cellMaster <- getColoredRange(sampleTable)) {
      getCellByPosition(compareTable, cellMaster.getRowIndex, cellMaster.getColumnIndex) match {
        case None              => Unit
        case Some(cellCompare) =>

          val (cellValueResult, cellValueMsg) = compareCellValues(cellMaster, cellCompare)

          val (cellFormulaResult, cellFormulaMsg) = compareCellFormulas(cellMaster, cellCompare)

          setCellComment(cellCompare, cellValueMsg + "\n" + cellFormulaMsg)

          val fontstyle = if (cellValueResult && cellFormulaResult) FontStyle.BOLD else FontStyle.ITALIC

          val font = new Font(fontName, fontstyle, fontSize)

          setCellStyle(cellCompare, font, cellValueResult && cellFormulaResult)
      }
    }
  }

  override def getCellByPosition(table: Table, column: Int, row: Int): Option[Cell] = Option(table.getCellByPosition(column, row))

  override def getColoredRange(master: Table): List[Cell] =
    (for {row <- 0 until maxRow
          column <- 0 until maxColumn
          oCell = master.getRowByIndex(row).getCellByIndex(column)
          if oCell.getCellBackgroundColorString != colorWhite} yield oCell).toList


  override def compareSheetConditionalFormatting(master: Table, compare: Table): List[String] = List.empty

  override def getSheetByIndex(document: SpreadsheetDocument, sheetIndex: Int): Table = document.getSheetByIndex(sheetIndex)

  override def getSheetCount(document: SpreadsheetDocument): Int = document.getSheetCount

  override def loadDocument(path: Path): Try[SpreadsheetDocument] = Try(SpreadsheetDocument.loadDocument(path.toFile))

  override def saveCorrectedSpreadsheet(compareDocument: SpreadsheetDocument, testPath: Path): Try[Path] = {
    val tPath = testPath.toString
    val fileNameNew = GFiles.getNameWithoutExtension(tPath) + CORRECTION_ADD_STRING + "." + GFiles.getFileExtension(tPath)
    val savePath = Paths.get(testPath.getParent.toString, fileNameNew)
    Try({
      if (!savePath.getParent.toFile.exists())
        Files.createDirectories(savePath.getParent)

      val fileOut: FileOutputStream = new FileOutputStream(savePath.toFile)
      compareDocument.save(fileOut)
      fileOut.close()

      savePath
    })
  }

  override def setCellComment(cell: Cell, message: String): Unit = if (message != null && message.nonEmpty) {
    cell.setNoteText(message)
  }

  override def setCellStyle(cell: Cell, font: Font, success: Boolean): Unit = {
    font.setColor(if (success) green else red)
    cell.setFont(font)
  }

}
